File: Solutions/Neetcode/ArraysAndHashing/6-EncodeAndDecodeStrings/solution.ts
const CHAR_SEPARATOR = "-";
const WORD_SEPARATOR = "_";

/**
 * Encodes a list of strings to a single string.
 */
function encode(strs: string[]): string {
  let encodedStrs = "";
  for (let i = 0; i < strs.length; i++) {
    if (i !== 0) encodedStrs += WORD_SEPARATOR;

    const word = strs[i];
    let encodedWord = "";

    for (let i = 0; i < word.length; i++) {
      if (i !== 0) encodedWord += CHAR_SEPARATOR;

      const charCode = word[i].charCodeAt(0);
      encodedWord += charCode;
    }

    encodedStrs += encodedWord;
  }

  return encodedStrs;
}

/**
 * Decodes a single string to a list of strings.
 */
function decode(s: string): string[] {
  let decodedStrs: string[] = [];

  for (let encodedWord of s.split(WORD_SEPARATOR)) {
    let decodedWord = "";
    for (let encodedChar of encodedWord.split(CHAR_SEPARATOR)) {
      if (!encodedChar) continue;
      decodedWord += String.fromCharCode(Number(encodedChar));
    }
    decodedStrs.push(decodedWord);
  }

  return decodedStrs;
}

/**
 * Your functions will be called as such:
 * decode(encode(strs));
 */

File: Solutions/Neetcode/ArraysAndHashing/6-EncodeAndDecodeStrings/solution.py
from typing import List


class Codec:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string."""
        encoded = ""
        for word in strs:
            encoded += str(len(word)) + "#" + word
        return encoded

    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings."""
        decoded = []
        i = 0
        while i < len(s):
            j = i
            while s[j] != "#":
                j += 1

            word_len = int(s[i:j])
            word_start = j + 1
            word_end = word_start + word_len
            word = s[word_start:word_end]
            decoded.append(word)
            i = word_end

        return decoded
